### ISSUE TITLE
Sometimes, kafka_exporter cannot connect to broker and generate too many sockets.

### ISSUE THREAD

**Issue by skonmeme** [danielqsj/kafka_exporter#54]
Usually it works well. 

Sometimes it could not connect to brokers and generates too many sockets (up to about 1,000 sockets open) and generate following error messages.

Whenever I restart kafka_exporter, it works well again.

---
Jul 20 12:36:38 172.27.115.150 kafka_exporter[20251]: time="2018-07-20T12:36:38+09:00" level=error msg="Can't get current offset of topic test partition %!s(int32=10): kafka: broker not connected" source="kafka_exporter.go:271"
...
---
![screen shot 2018-07-20 at 12 42 53 pm](https://user-images.githubusercontent.com/15389004/42982070-8a486960-8c1a-11e8-891b-8d1036e67578.png)

**Comment by dcarrier** [comment]
+1 we are seeing the same thing in our cluster.

**Comment by sirpkt** [comment]
+1

**Comment by jpbelanger-mtl** [comment]
We had the same issue after 87days of uptime. Restarting fixed it... during that time, metrics occasionally  failed. It got worst over time until it starting triggering out alerts.

![image](https://user-images.githubusercontent.com/18267109/44683997-dcc91180-aa15-11e8-85aa-5e1712c97e85.png)

**Comment by agolomoodysaada** [comment]
I would also like to add that not only do we get "Cannot get current offset" but also "Cannot get describe groups"

**Comment by eapache** [comment]
I'm not sure what I'm supposed to know about this app specifically but it looks like they're already calling `brokerOpen`: https://github.com/danielqsj/kafka_exporter/blob/master/kafka_exporter.go#L353

**Comment by agolomoodysaada** [comment]
Thanks @eapache for taking a look.
The exporter is basically an HTTP endpoint that gets called every interval (say 10s) and returns metrics from Kafka. In this case, on each HTTP request, we loop over each consumer group from __consumer_offsets and calculate retrieve different metrics from there. There appears to be a socket leak in this loop or some random disconnects in the client.

Is it better to create a single client for each loop iteration? or share the client across all iterations?
Are we missing anything in how the client is being used?

**Comment by eapache** [comment]
The code as written looks correct. It's possible there's something I'm overlooking, or it's possible you've managed to uncover a really subtle concurrency bug in Sarama. What does the race-detector say? Is anything interesting coming from the Sarama logs?

**Comment by hukaixuan** [comment]
I have the same issue,  found the reason from kafka server log:
`ERROR Closing socket for ... because of error (kafka.network.Processor)
kafka.network.InvalidRequestException: Error getting request for apiKey: 2 and apiVersion: 1`
And found that the kafka version is 0.10.0.0 but the kafka_exporter deafault kafka version is 1.0.0. Fix it by add --kafka.version=0.10.0.0

**Comment by agolomoodysaada** [comment]
@hukaixuan We still experience it on kafka 1.0.0. So it's still a bug. Also, it temporarily seems resolved after restarting the process. After a few hours, I imagine the symptoms will appear again.

**Comment by bojleros** [comment]
Hi,

It seems that kafka exporter is having a problem with metadata refresh:

![image](https://user-images.githubusercontent.com/26764966/49355473-0f933d80-f6c8-11e8-9def-1de06dda2448.png)

Things are always the same for me. I do get holes in my timeseries when meta refresh kicks in. There was an suggestion to create metadata cache. I can also sugest to get meta asynchronously, in a separate goroutine. Please have a look at my logs, it is from our sandbox so please let me know if i can help you troubleshooting.

(1h timestamp difference)
```
[sarama] 2018/12/03 05:10:17 Connected to broker at broker0:9092 (registered as #555)
[sarama] 2018/12/03 05:10:17 Connected to broker at broker1:9092 (registered as #556)
[sarama] 2018/12/03 05:10:17 Connected to broker at broker2:9092 (registered as #557)
[sarama] 2018/12/03 05:10:17 Closed connection to broker broker0:9092
[sarama] 2018/12/03 05:10:17 Closed connection to broker broker2:9092
[sarama] 2018/12/03 05:10:17 Closed connection to broker broker1:9092
[sarama] 2018/12/03 05:11:16 client/metadata fetching metadata for all topics from broker broker2:9092
[sarama] 2018/12/03 05:11:17 Connected to broker at broker1:9092 (registered as #556)
[sarama] 2018/12/03 05:11:17 Connected to broker at broker0:9092 (registered as #555)
[sarama] 2018/12/03 05:11:46 Failed to connect to broker broker2:9092: dial tcp 172.20.2.169:9092: i/o timeout
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [_one_of_our_topics_] from broker broker2:9092
[sarama] 2018/12/03 05:11:46 client/metadata fetching metadata for [__consumer_offsets] from broker broker2:9092
[sarama] 2018/12/03 05:11:47 Connected to broker at broker2:9092 (registered as #557)
[sarama] 2018/12/03 05:11:47 Closed connection to broker broker0:9092
[sarama] 2018/12/03 05:11:47 Closed connection to broker broker2:9092
[sarama] 2018/12/03 05:11:47 Closed connection to broker broker1:9092
[sarama] 2018/12/03 05:12:16 client/metadata fetching metadata for all topics from broker broker2:9092
[sarama] 2018/12/03 05:12:17 Connected to broker at broker1:9092 (registered as #556)
[sarama] 2018/12/03 05:12:17 Connected to broker at broker2:9092 (registered as #557)
[sarama] 2018/12/03 05:12:17 Connected to broker at broker0:9092 (registered as #555)
[sarama] 2018/12/03 05:12:17 Closed connection to broker broker0:9092
[sarama] 2018/12/03 05:12:17 Closed connection to broker broker2:9092
[sarama] 2018/12/03 05:12:17 Closed connection to broker broker1:9092
[sarama] 2018/12/03 05:13:16 client/metadata fetching metadata for all topics from broker broker2:9092
[sarama] 2018/12/03 05:13:17 Connected to broker at broker0:9092 (registered as #555)
[sarama] 2018/12/03 05:13:17 Connected to broker at broker2:9092 (registered as #557)
[sarama] 2018/12/03 05:13:17 Connected to broker at broker1:9092 (registered as #556)
[sarama] 2018/12/03 05:13:17 Closed connection to broker broker0:9092
[sarama] 2018/12/03 05:13:17 Closed connection to broker broker2:9092
[sarama] 2018/12/03 05:13:17 Closed connection to broker broker1:9092
[sarama] 2018/12/03 05:14:16 client/metadata fetching metadata for all topics from broker broker2:9092
[sarama] 2018/12/03 05:14:17 Connected to broker at broker0:9092 (registered as #555)
[sarama] 2018/12/03 05:14:17 Connected to broker at broker1:9092 (registered as #556)
[sarama] 2018/12/03 05:14:17 Connected to broker at broker2:9092 (registered as #557)
[sarama] 2018/12/03 05:14:17 Closed connection to broker broker2:9092
[sarama] 2018/12/03 05:14:17 Closed connection to broker broker0:9092
[sarama] 2018/12/03 05:14:17 Closed connection to broker broker1:9092
[sarama] 2018/12/03 05:15:16 client/metadata fetching metadata for all topics from broker broker2:9092
[sarama] 2018/12/03 05:15:17 Connected to broker at broker1:9092 (registered as #556)
[sarama] 2018/12/03 05:15:17 Connected to broker at broker0:9092 (registered as #555)
[sarama] 2018/12/03 05:15:17 Connected to broker at broker2:9092 (registered as #557)
[sarama] 2018/12/03 05:15:17 Closed connection to broker broker0:9092
[sarama] 2018/12/03 05:15:17 Closed connection to broker broker1:9092
[sarama] 2018/12/03 05:15:17 Closed connection to broker broker2:9092
[sarama] 2018/12/03 05:16:16 client/metadata fetching metadata for all topics from broker broker2:9092
[sarama] 2018/12/03 05:16:17 Connected to broker at broker1:9092 (registered as #556)
[sarama] 2018/12/03 05:16:17 Connected to broker at broker0:9092 (registered as #555)
[sarama] 2018/12/03 05:16:17 Connected to broker at broker2:9092 (registered as #557)
[sarama] 2018/12/03 05:16:17 Closed connection to broker broker0:9092
[sarama] 2018/12/03 05:16:17 Closed connection to broker broker2:9092
[sarama] 2018/12/03 05:16:17 Closed connection to broker broker1:9092
```

**Comment by bojleros** [comment]
BTW, Why it is always fetching meta from the same broker ?

**Comment by bojleros** [comment]
Hi guys ! Any news ? For us it is a showstoper :(

**Comment by bojleros** [comment]
Hi ! Did anyone tried if this PR helps for this issue ?  I do really think it is all around metadata handling:

https://github.com/danielqsj/kafka_exporter/pull/75

@danielqsj Would you please give us at least short answer ?

**Comment by danielqsj** [comment]
@agolomoodysaada @bojleros  #75 merged, can you try the newest image and see whether things are getting better?

**Comment by agolomoodysaada** [comment]
Deployed it. Will test for a few days and report back.

**Comment by bojleros** [comment]
Sure , I'll post my results as soon as we have one.

**Comment by bojleros** [comment]
@danielqsj @jorgelbg @agolomoodysaada 
Hi , hhat i see now is a metadata refresh every 30s and that's fine however i do still get gaps from time to time.

```
[sarama] 2019/02/01 10:42:26 client/metadata fetching metadata for all topics from broker event-sea-2:9092
time="2019-02-01T10:42:41Z" level=info msg="Refreshing client metadata" source="kafka_exporter.go:233"
[sarama] 2019/02/01 10:42:41 client/metadata fetching metadata for all topics from broker event-sea-2:9092
[sarama] 2019/02/01 10:42:41 Failed to connect to broker event-sea-2:9092: dial tcp 172.a.a.a:9092: i/o timeout
time="2019-02-01T10:42:41Z" level=error msg="Cannot get current offset of topic aaaa partition 0: dial tcp 172.a.a.a:9092: i/o timeout" source="kafka_exporter.go:274"
time="2019-02-01T10:42:41Z" level=error msg="Cannot get current offset of topic aaaa partition 0: dial tcp 172.a.a.a:9092: i/o timeout" source="kafka_exporter.go:274"
time="2019-02-01T10:42:41Z" level=error msg="Cannot get current offset of topic aaaa partition 0: dial tcp 172.a.a.a:9092: i/o timeout" source="kafka_exporter.go:274"
time="2019-02-01T10:42:41Z" level=error msg="Cannot get current offset of topic aaaa partition 0: dial tcp 172.a.a.a:9092: i/o timeout" source="kafka_exporter.go:274"
[sarama] 2019/02/01 10:42:41 Connected to broker at event-sea-2:9092 (registered as #102)
[sarama] 2019/02/01 10:42:41 Closed connection to broker event-sea-0:9092
[sarama] 2019/02/01 10:42:41 Closed connection to broker event-sea-2:9092
[sarama] 2019/02/01 10:42:41 Closed connection to broker event-sea-1:9092
[sarama] 2019/02/01 10:42:41 Connected to broker at event-sea-2:9092 (registered as #102)
[sarama] 2019/02/01 10:42:41 Connected to broker at event-sea-0:9092 (registered as #100)
[sarama] 2019/02/01 10:42:41 Connected to broker at event-sea-1:9092 (registered as #101)
[sarama] 2019/02/01 10:42:41 Closed connection to broker event-sea-1:9092
[sarama] 2019/02/01 10:42:41 Closed connection to broker event-sea-2:9092
[sarama] 2019/02/01 10:42:41 Closed connection to broker event-sea-0:9092
[sarama] 2019/02/01 10:42:56 client/metadata fetching metadata for all topics from broker event-sea-2:9092```

Other thing is this i/o timeout and it's handling. I think such error can happen but kafka_exporter should be able to make at least 1 or 2 fast retries (to the one of different brokers) before updating it's metadata and generating a gap. Another finding is that kafka_exporter keeps polling only one of brokers for metadata change instead of making a round-robin across all brokers.

**Comment by bojleros** [comment]
Hi, We have installed kafka-exporter on prod just to get additional metrics without alerting. Bad news is that i do still see occassional gaps in my metrics. @agolomoodysaada how does it work for you ? Did you changed metadata refresh interval ?

**Comment by alexey-dushechkin** [comment]
Not sure if it's the right place, but I observe something similar after upgrading Prometheus from 2.5.0 to 2.7.1. I have odd setup where single kafka 1.1.1 exporter 1.2.0 is scraped by 6 servers simultaneously, and after update 30-40% of kafka_consumergroup_lag metric values are missing depending on RTT between Prometheus and exporter (I have 0.5, 50 and 90 ms).

And my guess it's a race condition, because commenting the [defer broker.Close()](https://github.com/danielqsj/kafka_exporter/blob/v1.2.0/kafka_exporter.go#L357) fixes the issue right away.

**Comment by jorgelbg** [comment]
@alexey-dushechkin I think I've run into the same conclusion.

I'm running the Kafka exporter as well with multiple Prometheus scraping the target (4 in my case). After enabling the extended logging for sarama I see a lot of consecutive connections&disconnections (from the brokers) on the logs, within the same second:

```
[sarama] 2019/02/11 09:33:59 Connected to broker at kafka19:9092 (registered as #19)
[sarama] 2019/02/11 09:33:59 Connected to broker at kafka0:9092 (registered as #0)
[sarama] 2019/02/11 09:33:59 Connected to broker at kafka4:9092 (registered as #4)
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka3:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka4:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka18:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka19:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka10:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka8:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka9:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka7:9092
[sarama] 2019/02/11 09:33:59 Connected to broker at kafka10:9092 (registered as #10)
[sarama] 2019/02/11 09:33:59 Connected to broker at kafka8:9092 (registered as #8)
[sarama] 2019/02/11 09:33:59 Connected to broker at kafka19:9092 (registered as #19)
[sarama] 2019/02/11 09:33:59 Connected to broker at kafka9:9092 (registered as #9)
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka10:9092
[sarama] 2019/02/11 09:33:59 Closed connection to broker kafka8:9092
```

Also at the same time on the normal logs I see the following:

```
time="2019-02-11T09:34:17Z" level=error msg="Cannot get consumer group: kafka: broker not connected" source="kafka_exporter.go:372"
```
This means that from the call to [`broker.Open()`](https://github.com/danielqsj/kafka_exporter/blob/master/kafka_exporter.go#L362) on and the call to [`broker.ListGroups()`](https://github.com/danielqsj/kafka_exporter/blob/master/kafka_exporter.go#L368) the connection was closed. 

In theory we don't need to disconnect after each request to the broker. Although Kafka has the `connections.max.idle.ms` on recent versions (defaults to 10m, I think), the connection is going to be used long before that (unless you have a `scrape_interval` set to `>10m`). 

I've deployed a patched version this morning and will post results here tomorrow.

**Comment by bojleros** [comment]
Currently i do only have a single prometheus talking with multiple kafka-exporters so for me it should not be a concurrency problem.

> In theory we don't need to disconnect after each request to the broker. Although Kafka has the `connections.max.idle.ms` on recent versions (defaults to 10m, I think), the connection is going to be used long before that (unless you have a `scrape_interval` set to `>10m`).

Yes . There is a such parameter. I do not remember exact kafka version introducing it . I do only remember that our versions are newer ones but i do also see some stalled partition connections (10m+) from time to time so i am not absolutely sure it is working correctly.

**Comment by jorgelbg** [comment]
@bojleros From your logs it got my attention that your exporter is actually timing out when connecting to the broker:

```
[sarama] 2019/02/01 10:42:41 Failed to connect to broker event-sea-2:9092: dial tcp 172.a.a.a:9092: i/o timeout
```

Also is what you get directly from the exporter:

```
time="2019-02-01T10:42:41Z" level=error msg="Cannot get current offset of topic aaaa partition 0: dial tcp 172.a.a.a:9092: i/o timeout" source="kafka_exporter.go:274"
```

This looks like it really can't connect to that particular broker. Could you try to pass multiple brokers at startup time? Something like:

```
--kafka.server=kafka1:9092 --kafka.server=kafka2:9092
```

I've tested locally by passing 1 existing broker and one dummy broker and it seems to work. The only issue that I've encountered is that it will try again on each request to connect to the non-healthy broker first (if it was selected initially as the "main" broker). This retry mechanism is provided by sarama directly. But you need to pass multiple brokers in the command line.

**Comment by bojleros** [comment]
@jorgelbg Thats how my commandline looks like. Every single broker in my cluster is on that list. Since my clusters are rather small i do include every broker on the list. I to not consider it like it was a '--bootstrap-server' known from kafka tools.

```
    spec:
      containers:
      - args:
        - --log.enable-sarama
        - --log.level=debug
        - --kafka.version=1.1.0
        - --kafka.server=kafka-sea-0:9092
        - --kafka.server=kafka-sea-1:9092
        - --kafka.server=kafka-sea-2:9092
        - --kafka.server=kafka-sea-3:9092
        - --kafka.server=kafka-sea-4:9092
        - --kafka.server=kafka-sea-5:9092
```

**Comment by jorgelbg** [comment]
I've been running a patched version without the `broker.Close()` (after the broker is consumed) for a couple of days. On the metrics side it looks a lot better than before:

![image](https://user-images.githubusercontent.com/1291846/52719533-18e13600-2fa6-11e9-8fbc-931b59107504.png)

We have less gaps now. I've found ocasional disconnections to the broker that I'm connecting to but it reconnected automatically after that and no gap was present at the same time.

Memory usage is more stable now (as you can see) and although it has been increasing. This job was running with plenty of RAM (200MB) so perhaps it was not triggering any GC actions. I've deployed a new job with 100MB of RAM, will leave it running for some time and will check later on.

It would be interesting to see if the RAM usage of your exporter follows the same pattern @alexey-dushechkin. 

![image](https://user-images.githubusercontent.com/1291846/52731247-8b114500-2fbd-11e9-97b5-ce57c168f76b.png)

**Comment by bojleros** [comment]
My limits are following:

```
    Limits:
      memory:  256Mi
    Requests:
      cpu:        50m
      memory:     32Mi
```

![image](https://user-images.githubusercontent.com/26764966/52766778-9057a980-3028-11e9-8aa8-83da30d395e8.png)

There seems to be a plenty of ram and i see no OOM at all. I do use following to generate this graph:

`sum (container_memory_working_set_bytes{pod_name=~"kafka-exporter.*"}) by (pod_name)`

It is possible that memory usage is related to the broker count , number of topics , etc but first, are we using the same metrics ? How many brokers do you run @jorgelbg  ?

PS. My kafka_exporter runs on k8s cluster but i also have node_exporters on each kafka broker. No gaps on node_exporter metrics at all.

**Comment by jorgelbg** [comment]
@bojleros In my case, it's a bit different I'm deploying in a Nomad cluster. I'm monitoring the RSS memory for the allocation that I get from the Nomad telemetry. I think the equivalent to Kubernetes would be `container_memory_rss`, probably.

Currently I'm connecting to 12 brokers with around 450 topics. So yes, I think that memory consumtion is related to this. I was a bit concerned about the increasing trend that I saw on my graphs. But It has been running fine with 100MB of RAM since ysterday, which is 1/2 of what it had before. 

I'll take a look at kage as well.

**Comment by bojleros** [comment]
`sum (container_memory_rss{pod_name=~"kafka-exporter.*"}) by (pod_name)`

![image](https://user-images.githubusercontent.com/26764966/52783329-dd03aa80-3051-11e9-92e3-5027c90d19d1.png)

My scale is 6 brokers * ~15 topics * 12 partitions/topic * 2 replicas. Some of topics are compacted but that should not be an issue for monitoring. Looking at cpu usage (k8s metrics) i see some spikes around 10 milicores with average much more lower. No evidence for resource starvation however i am going to run this exporter outside of k8s and let you know.

**Comment by alexey-dushechkin** [comment]
@jorgelbg I have some oranges to compare to your apples, because all I have is memory metrics from the exporter itself (from single instance, other looks alike):
![kafka_exporter_memory](https://user-images.githubusercontent.com/20111878/52792252-7f954b00-309d-11e9-89a5-37206acd2538.png)
Prometheus was fully upgraded on Feb 8th (before that only 1 out of 6 servers was updated), patched exporter was deployed late Feb 11 (you might see bold line at the bottom changes color).

I don't experience any holes in the metrics since then, but my setup is pretty simple: single Kafka server per exporter, single topic, 2 partitions. And I think extra broker.Close() should be filled as a separate issue because it seems different from what others experience here.

**Comment by bojleros** [comment]
Yesterday i have found that our Prometheus is configured with scrape_interval: 1m. I have configured kafka job back to 15s since i feel 1m can amplify gap effects. I am going to configure additional exporter outside k8s cluster.

@alexey-dushechkin Are you sure that having one exporter dedicated to each broker is a correct setup ?  As far as I understand how sarama and kafka protocol works it is expected that each kafka_exporter will try to contact every broker in cluster even if you use only one broker in command line. Kafka protocol is propagating cluster node list directly to the clients and one passed in command line act as initial (bootstrap) broker. Am i right ?

**Comment by bojleros** [comment]
@jorgelbg , I am now making a rolling restart of one of our clusters. I do still get gaps with every cluster topology change. This is what i see while trying to connect exporter directly when gaps occur:

```
$ curl -v localhost:9308
* Rebuilt URL to: localhost:9308/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 9308 (#0)
> GET / HTTP/1.1
> Host: localhost:9308
> User-Agent: curl/7.61.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 15 Feb 2019 12:14:59 GMT
< Content-Length: 188
< Content-Type: text/html; charset=utf-8
< 
<html>
	        <head><title>Kafka Exporter</title></head>
	        <body>
	        <h1>Kafka Exporter</h1>
	        <p><a href='/metrics'>Metrics</a></p>
	        </body>
* Connection #0 to host localhost left intact
$ curl -v localhost:9308/metrics
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 9308 (#0)
> GET /metrics HTTP/1.1
> Host: localhost:9308
> User-Agent: curl/7.61.1
> Accept: */*
> 
^C
```

No answer from endpoint `/metrics` at all.

**Comment by bojleros** [comment]
I have finally managed to install kafka_exporter on one of kafka nodes. It is how it looks like (blue vs orange):

![image](https://user-images.githubusercontent.com/26764966/52954488-a3e97400-338a-11e9-9642-9838751e57f3.png)

It is initial result so i recognize it only as a weak proof of troubles somewhere between kafka_exporter and kafka brokers itself. Gonna post if it changes.

**Comment by jorgelbg** [comment]
@bojleros From the output of your `curl` command (confirmed by the graph) seems like there is something a real connectivity issue between the exporter and the Kafka broker.

@alexey-dushechkin thanks for the info. My exporter instance is using ~50MB of RAM normally, and I don't see any gaps in the data either. 👍  with putting this change in a different PR. 

Perhaps the `broker.Close()` was related to the main issue reported here. Since the connection was  opened/closed many times. In the meantime we updated to a newer sarama version, so that could have been also related. Anyhow that is hard to probe now unless the original reporters give a try to the latest version.

**Comment by bojleros** [comment]
@jorgelbg, Yes , there is something bad happening between exporter and kafka cluster. Currently i have kafka_exporter and kage running in parallel (in a same k8s cluster and namespace) monitoring same clusters. No gaps for kage so far.

On the other side i have installed kafka_exporter directly on first node of each cluster i have. They are working correctly (no gaps) even with `--kafka.server localhost:9092`. Lack of monitoring HA makes this sollution obviously unacceptable.

In friday i was stress-testing my stg clusters and i saw negative lag reported by kafka_exporter (old issue i believe). I hope kafka_exporter gets more mature soon.

**Comment by claudio-benfatto** [comment]
Hi, we are using the kafka exporter in production for several of our Kafka cluster and unfortunately hitting this problem as well. Any news about a possible solution? Any help needed?

Thanks!

**Comment by jorgelbg** [comment]
@claudio-benfatto are you seeing the high number of open sockets or the gaps in the data?

**Comment by claudio-benfatto** [comment]
Hi @jorgelbg we are having problems with the file descriptors due to the high number of open sockets:

ie.
```
# ls -l /proc/5716/fd |wc -l
1025 
```
The file descriptors are open to a socket.
```
# lsof -nPi :9092 |grep 5716 |awk '{print $10}' |sort |uniq -c |sort
    186 (ESTABLISHED)
    832 (CLOSE_WAIT)
```

Thanks!

**Comment by jorgelbg** [comment]
Which Kafa version are you running? I haven't had that issue myself. 
Would you mind giving PR#90 a try? It avoids closing the Kafka connection after each scrape.

**Comment by claudio-benfatto** [comment]
I'm running Kafka `2.0.0`.
I can give it a try, however I would be more inclined to do it if it was merged and released :)

**Comment by jorgelbg** [comment]
@claudio-benfatto unfortunately, merging it is out of my hands 🙂. Do you use Docker by any chance? I could push the image that we maintain in our internal registry to docker hub. It contains PR #90 and #93.

**Comment by claudio-benfatto** [comment]
@jorgelbg thank you for the offer ;) No, we package the exporter in a rpm. However do not worry, I can fork the repo and try the PRs you suggested

**Comment by claudio-benfatto** [comment]
anyhow debugging it a bit more, this is what I noticed:

`strace`
```
read(22, "\0\1\250\373\0\1\360^\0\0\1\2\0\\Public-DataPro-Ygg"..., 4096) = 4096
read(22, "sil-OmniProd-AmplitudeEvents-1\0\0"..., 104703) = 104703
futex(0xc420750948, FUTEX_WAKE, 1)      = 1
futex(0xc420596548, FUTEX_WAKE, 1)      = 1
futex(0xbf57e0, FUTEX_WAKE, 1)          = 1
futex(0xbf57c0, FUTEX_WAKE, 1)          = 1
write(22, "\0\0\2343\0\t\0\1\0\1\360_\0\16kafka_exporter\0002Du"..., 39991) = 39991
futex(0xc420596548, FUTEX_WAKE, 1)      = 1
futex(0xbf57e0, FUTEX_WAKE, 1)          = 1
futex(0xbf57c0, FUTEX_WAKE, 1)          = 1
futex(0xbf57e0, FUTEX_WAKE, 1)          = 1
futex(0xbf57c0, FUTEX_WAKE, 1)          = 1
read(22, 0xc420468000, 4096)            = -1 EAGAIN (Resource temporarily unavailable)
futex(0xbf2168, FUTEX_WAIT, 0, NULL)    = 0
pselect6(0, NULL, NULL, NULL, {0, 3000}, 0) = 0 (Timeout)
futex(0xc420596548, FUTEX_WAKE, 1)      = 1
futex(0xbf2168, FUTEX_WAIT, 0, NULL)    = 0
pselect6(0, NULL, NULL, NULL, {0, 3000}, 0) = 0 (Timeout)
futex(0xbf2168, FUTEX_WAIT, 0, NULL)    = 0
epoll_wait(4, {{EPOLLOUT, {u32=1289494152, u64=139806769946248}}}, 128, 0) = 1
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=1291142080, u64=139806771594176}}}, 128, -1) = 1
futex(0xbf1950, FUTEX_WAKE, 1)          = 1
futex(0xbf1870, FUTEX_WAKE, 1)          = 1
read(26, "\0\1\250\373\0\5W\335\0\0\1\2\0\\Public-DataPro-Ygg"..., 4096) = 4096
read(26, "sil-OmniProd-AmplitudeEvents-1\0\0"..., 104703) = 104703
futex(0xc420596548, FUTEX_WAKE, 1)      = 1
futex(0xc420750948, FUTEX_WAKE, 1)      = 1
futex(0xc420750948, FUTEX_WAKE, 1)      = 1
futex(0xbf5ce0, FUTEX_WAKE, 1)          = 1
futex(0xbf5cc0, FUTEX_WAKE, 1)          = 1
write(26, "\0\0\2348\0\t\0\1\0\5W\336\0\16kafka_exporter\0007Du"..., 39996) = 39996
futex(0xc420751148, FUTEX_WAKE, 1)      = 1
futex(0xbf5ce0, FUTEX_WAKE, 1)          = 1
futex(0xbf5cc0, FUTEX_WAKE, 1)          = 1
futex(0xbf5ce0, FUTEX_WAKE, 1)          = 1
futex(0xbf5cc0, FUTEX_WAKE, 1)          = 1
read(26, 0xc420a45000, 4096)            = -1 EAGAIN (Resource temporarily unavailable)
futex(0xbf2168, FUTEX_WAIT, 0, NULL)    = 0
epoll_wait(4, {{EPOLLOUT, {u32=1289937728, u64=139806770389824}}}, 128, 0) = 1
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=1290307880, u64=139806770759976}}}, 128, -1) = 1
futex(0xbf1950, FUTEX_WAKE, 1)          = 1
futex(0xbf1870, FUTEX_WAKE, 1)          = 1
```

many `CLOSE_WAIT` sockets
```
prometheu  6259 prometheus    7u  IPv4 3355355      0t0  TCP ip-10-231-64-125.eu-west-1.compute.internal:40942->ip-10-231-86-121.eu-west-1.compute.internal:XmlIpcRegSvc (CLOSE_WAIT)
```
and many routines running (one of them is the one I straced):
```
493       6259  6244  6259  0   52 Jul26 ?        00:11:16 /usr/local/bin/prometheus-kafka-exporter --log.level=error --log.format=logger:stderr?json=true --kafka.server=localhost:9092
```

**Comment by forget6** [comment]
+1 we are seeing the same thing in our cluster.i restart the kafka_exporter. kafka_exporter is ok now. but I don't know why. what happened?  

`time="2019-08-14T11:06:04+08:00" level=error msg="Cannot get offset of group test_topic: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:06:19+08:00" level=error msg="Cannot get offset of group testdb: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:06:20+08:00" level=error msg="Cannot get offset of group test_topic: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:07:19+08:00" level=error msg="Cannot get offset of group test_topic: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:07:34+08:00" level=error msg="Cannot get offset of group testdb: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:07:35+08:00" level=error msg="Cannot get offset of group test_topic: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:08:04+08:00" level=error msg="Cannot get describe groups: kafka: broker not connected" source="kafka_exporter.go:373"
time="2019-08-14T11:09:19+08:00" level=error msg="Cannot get describe groups: kafka: broker not connected" source="kafka_exporter.go:373"
time="2019-08-14T11:09:34+08:00" level=error msg="Cannot get offset of group testdb: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:10:19+08:00" level=error msg="Cannot get describe groups: kafka: broker not connected" source="kafka_exporter.go:373"
time="2019-08-14T11:10:34+08:00" level=error msg="Cannot get offset of group testdb: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:11:20+08:00" level=error msg="Cannot get offset of group testdb: kafka: broker not connected" source="kafka_exporter.go:384"
time="2019-08-14T11:11:34+08:00" level=error msg="Cannot get describe groups: kafka: broker not connected" source="kafka_exporter.go:373"`

`./kafka_exporter --web.listen-address=:9308 --kafka.server=xx.xx.xx.xxx:xxx --kafka.server=xx.xx.xx.xxx:xxx --sasl.enabled --sasl.username=xxxx --sasl.password=xxxxxx`

**Comment by jorgelbg** [comment]
@forget6 would you mind running the exporter with the debug log level? It would be interesting to see why it is failing to connect to the broker. Are you running master or a released version? Unfortunately, I haven't been able to reproduce this issue in my setup.

**Comment by jadams-gr8** [comment]
Maybe somewhat related, I recently tried to deploy kafka_exporter in a new environment and noticed that the metrics took a really long time to collect (in my case about 11 seconds with a dozen or so topics with the largest having 50 partitions). Trying to figure out why brought me to this issue. I took a dive into the exporter and learned that it was due to the latency between the exporter and the kafka cluster. 
In most of my deployments there is only a few ms between the exporter and kafka, but in this scenario there was about 75ms. Each partition's offsets need to be queried from kafka live. This is done once for the current offset and once for the oldest offset. This must be done for each partition of each topic, so if there are a large number of network round trips. 
I wound up forking the repo to try and solve this for my use case. I did some refactoring and was able to get more concurrency on the network calls and that brought my collection time down to 5 seconds. Adding a rudimentary connection pool (the sarama.Client instances process commands in order according to the docs) brought that down further to 3 seconds. That time is now split relatively equally between the topic metrics and the consumer group metrics. 3 seconds is good enough for my case so I stopped optimizing at that point. 
Not sure if the latency and serialized queries contribute to the scenarios of the other commenters, but it's probably worth a look. This exporter has the potential to create a lot of network packets especially as the number of topics and partitions increases.

**Comment by jorgelbg** [comment]
@jadams-gr8 I would be interested in looking are your changes, perhaps they can be merged. I understand that your latency times were very high probably due to specifics of your infrastructure or some other particularities but I would be interested in testing your changes. 

I'm running the current version and monitoring 5 different Kafka clusters (the biggest one with ~270 active topics). Right now in master, the connection to Kafka is closed after each scrape interval which caused some issues, see https://github.com/danielqsj/kafka_exporter/issues/54#issuecomment-463293595. 

Did you try with the PR #90  by any chance? If you can make your fork available I can give it a try and perhaps you can contribute your changes as a PR.

**Comment by sysadmind** [comment]
I have not tried the patch, but I don't believe that patch to be particularly relevant to the specific time issue that I had. My repo is available here: https://github.com/sysadmind/kafka_exporter/
I think my changes _may_ be of use to you especially if some of your topics have a lot of partitions because this adds more concurrency to reduce waiting on network latency. I'd be open to submitting my changes as a PR if you can validate that they work for you (maybe even with your patch on top of it). I didn't want to open a PR yet since I'm not sure my work won't break things for other users and this repo looks to be lacking much activity recently.
My latency is because the exporter is actually running in a different datacenter than my kafka cluster as it's a development environment and some things are transitioning between datacenters. I could have moved my exporter but my current prometheus deployment makes that harder than it sounds on the surface.

**Comment by arkyhuang** [comment]
Same issue under kafka_exporter 1.2.0 and kafka 2.2.0, but unfortunately, restart did not solve issue yet. Any update or release about it ?

**Comment by Amojow** [comment]
I had similar error on my kafka exporter (kafka exporter 1.2)
```
time="2020-01-30T10:58:03Z" level=error msg="Cannot get describe groups: kafka: broker not connected" source="kafka_exporter.go:373"
```
I used Strimzi chart to deploy the strimzi-kafka-operator:0.16.2 and then my kafka cluster. 
Kubernetes v1.14.3

Fixed it also by reloading the kafka exporter

**Comment by prasus** [comment]
Greetings, I am experiencing the same issues since last few weeks, the restart sometime solvest the problems, but it reoccurs intermittently leading to missing data in our Grafana dashboards. 

```
time="2020-08-12T04:06:47Z" level=error msg="Cannot get oldest offset of topic xxxxx partition 31: kafka: broker not connected" source="kafka_exporter.go:296"
time="2020-08-12T04:06:50Z" level=error msg="Cannot get current offset of topic xxxxx partition 129: kafka: broker not connected" source="kafka_exporter.go:284"
time="2020-08-12T04:37:15Z" level=error msg="Cannot get consumer group: kafka: broker not connected" source="kafka_exporter.go:361"
```
I also tried increasing the `scrape_timeout` value on the Prometheus service monitor, but it doesn't help much. 

Any news on the update/fix about this? Thanks! 🙏

**Comment by talonx** [comment]
I am seeing this too, with the latest exporter image. Any solutions or workarounds? This is the only Kafka exporter out there I could find - so it would be nice if this were fixed.

**Comment by ximis** [comment]
I found this too. when I started writing data into kafka , Kafka Exporter is not responsive。

**Comment by atrbgithub** [comment]
Also seeing this. 

https://github.com/Shopify/sarama/issues/1857#issue-767619575 may be related?

**Comment by alok87** [comment]
Adding one more `broker.Open()` fixed the issue.

I was already opening the broker connection before the function call.

```go
err = broker.Open(t.client.Config())
if err != nil && err != sarama.ErrAlreadyConnected {
return defaultLag, fmt.Errorf("Error opening broker connection again, err: %v", err)
}

k.consumerGroupLag(id, topic, partiton, broker)
```

```go
 func (t *kafkaWatch) consumerGroupLag(id string, topic string, partition int32, broker *sarama.Broker) {
        .... do something with client ....(t.client.GetOffset)
   
        // one more broker open needed
        err = broker.Open(t.client.Config())
	if err != nil && err != sarama.ErrAlreadyConnected {
		return defaultLag, fmt.Errorf("Error opening broker connection again, err: %v", err)
	}

	offsetFetchResponse, err := broker.FetchOffset(&offsetFetchRequest)
```

**Comment by alok87** [comment]
No this also did not work :( still geting not connected broker.

**Comment by alesj** [comment]
@alok87 did you try with the latest release?

Here I've merged some other perf PR, plus added a way to limit getTopicMetrics routines
* https://github.com/alesj/kafka_exporter/tree/fork1
... if it helps ...

**Comment by alok87** [comment]
@alesj Yes using v1.3.0 still troubled with 
```
time="2021-04-21T15:16:22Z" level=error msg="Cannot get offset of group CG: kafka: broker not connected" source="kafka_exporter.go:412"
```

Trying your fork!

**Comment by alok87** [comment]
Deployed your optimized fork https://github.com/alesj/kafka_exporter/tree/fork1
Here is the image for it: [practodev/kafka-exporter:fork1](https://hub.docker.com/repository/docker/practodev/kafka-exporter/tags?page=1&ordering=last_updated)
We also have a big cluster, 1000s of consumer groups! 

This optimization seems to work!

@danielqsj please take this PR in, for large cluster support!

**Comment by atrbgithub** [comment]
@alesj Your fork has resolved our issue where the exporter's memory usage would sometimes spike resulting in an oom. Thank you! Memory usage is now much lower and stays low.

**Comment by alok87** [comment]
@danielqsj any plan to take this in? big clusters with many consumer groups the master of this repo does not work.

**Comment by danielqsj** [comment]
Cherry-picked from https://github.com/danielqsj/kafka_exporter/pull/228, and released v1.4.1. @alok87 and @atrbgithub and @all, please have a try.

**Comment by alok87** [comment]
Thanks @danielqsj 
I have deployed `danielqsj/kafka-exporter:v1.4.1`. I'll Let you know how it goes.

**Comment by alok87** [comment]
Update: it's working in 1000+ consumer group cluster in prod.  I will update if any issue happens. It should not. Thanks @danielqsj and @skonmeme

**Comment by alok87** [comment]
Update: Had to revert v1.4.1 the target endpoint became very slow to respond after this, and target was down.
![Screenshot 2021-09-08 at 4 26 08 PM](https://user-images.githubusercontent.com/9006763/132497460-94176268-1bd6-40e1-857c-91759972c09f.png)

@danielqsj
